dipy/reconst/ivim.py

```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/python
+""" Classes and functions for fitting ivim model """
+from __future__ import division, print_function, absolute_import
+import warnings
+import functools
+import numpy as np
+from scipy.optimize import minimize
+
+from dipy.core.gradients import gradient_table
+from dipy.reconst.base import ReconstModel
+from dipy.reconst.dti import _min_positive_signal
+
+
+def ivim_function(params, bvals):
+    """The ivim function
+    Parameters
+    ----------
+    params : array
+             The IVIM parameters [S0, f, D_star, D]
+
+    bvals  : array
+             bvalues
+    """
+    S0, f, D_star, D = params
+    S = S0 * (f * np.exp(-bvals * D_star) + (1 - f) * np.exp(-bvals * D))
+    return S
+
+
+def _ivim_error(params, bvals, signal):
+    """Error function to be used in fitting the model
+    """
+    return np.linalg.norm(signal - ivim_function(params, bvals))
```

Comment:
Yes: and note also that if you are using `leastsq` the error function
should return the errors (because it calculates the square separately!) and
if you are using `minimize` you want to give it the already squared SSE.

On Sat, Jun 4, 2016 at 2:04 PM, Eric Peterson <notifications@github.com>
wrote:

> In dipy/reconst/ivim.py
> <https://github.com/nipy/dipy/pull/1058#discussion_r65807651>:
>
> > +    ----------
> > +    params : array
> > +             The IVIM parameters [S0, f, D_star, D]
> > +
> > +    bvals  : array
> > +             bvalues
> > +    """
> > +    S0, f, D_star, D = params
> > +    S = S0 * (f * np.exp(-bvals * D_star) + (1 - f) * np.exp(-bvals * D))
> > +    return S
> > +
> > +
> > +def _ivim_error(params, bvals, signal):
> > +    """Error function to be used in fitting the model
> > +    """
> > +    return np.linalg.norm(signal - ivim_function(params, bvals))
>
> Actually this should be changed to (signal - ivim_function(params,
> bvals))**2. This is the traditional sum of squares error, the square root
> that the norm is doing may be the cause of your ills.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1058/files/3f237344e82c0beaefd135ecd54fe3d77fd23c8e#r65807651>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAHPNiVt0QINl4e6N5kC_aXzE1Mvb2UBks5qIehEgaJpZM4Imo2R>
> .
>